File: rpi/pseudoSPI_speedtest.py
# code for min/max communication with prop.spin
import RPi.GPIO as GPIO
import numpy as np

mosi = 11
miso = 12
mclk = 13
sclk = 15

def setup():
    GPIO.setmode(GPIO.BOARD)
    GPIO.setwarnings(False)
    GPIO.setup(mclk,GPIO.OUT, initial=GPIO.LOW)
    GPIO.setup(mosi,GPIO.OUT)
    GPIO.setup(sclk,GPIO.IN)
    GPIO.setup(miso,GPIO.IN)

def shutdown():
    GPIO.cleanup()

def IO1(dataout):
    GPIO.output(mclk, GPIO.HIGH)
    while GPIO.input(sclk) == 0:
        continue

    if dataout == 1:
        GPIO.output(mosi,GPIO.HIGH)
    else:
        GPIO.output(mosi,GPIO.LOW)
    GPIO.output(mclk, GPIO.LOW)

    while GPIO.input(sclk) == 1:
        continue

    datain = GPIO.input(miso)

    return datain
        
def IO16(dataout):
    n = 16
    i = 0
    datain = 0
    
    while i < n:
        datain = datain | IO1(dataout & 1)<<i

        dataout = dataout >> 1
        
        i = i + 1

    return datain

def runMe():
    setup()
    out = 15
    for i in xrange(10000):
        inp = IO16(out)
        #print("sent",out,"got",inp)
    shutdown()

runMe()

File: rpi/ccode/pseudoSPI_speedtest.c
#include <stdio.h>
#include <wiringPi.h>
#include <stdlib.h>

#define MOSI 0
#define MISO 1
#define MCLK 2
#define SCLK 3

void setup(){
  if(wiringPiSetup() == -1){
    printf("wiringPiSetup failed...?\n");
    exit(1);
  }

  pinMode(MCLK, OUTPUT);
  pinMode(SCLK, INPUT);
  pinMode(MOSI, OUTPUT);
  pinMode(MISO, INPUT);
}

void shutdown(){
  pinMode(MCLK, INPUT);
  pinMode(SCLK, INPUT);
  pinMode(MOSI, INPUT);
  pinMode(MISO, INPUT);
}

int io1(int send){
  int in;

  digitalWrite(MCLK,1);

  while(digitalRead(SCLK)==0){
    continue; // add a delay here?
    //delay(1); // 1 ms delay
  }

  digitalWrite(MOSI,send);
  digitalWrite(MCLK,0);

  while(digitalRead(SCLK)==1){
    continue; // add a delay here?
    //delay(1); // 1 ms delay
  }

  in = digitalRead(MISO);

  return in;
}

int io16(int send){
  int i=0;
  int in=0;
  
  for(i=0; i<16; ++i){
    in = in | (io1(send & 1) << i);
    send = send >> 1;
  }

  return in;
}

int main(void){
  int in,out,i;
  setup();

  printf("starting up...\n");
  out = 15;
  for(i=0; i<10000; ++i){
    in = io16(out);
  }
  printf("finished!\n");

  shutdown();

  return 0;
}
